src/lib/vasi-sync/src/sync.rs

```diff
@@ -30,86 +30,158 @@ loom::lazy_static! {
-    //
-    // TODO: Consider using
-    // [`as_mut_ptr`](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU32.html#method.as_mut_ptr)
-    // here once it's stabilized.
-    type SourceT = AtomicU32;
-    type TargetT = u32;
-    let futex_word: &SourceT = futex_word;
-    static_assertions::assert_eq_size!(SourceT, TargetT);
-    static_assertions::assert_eq_align!(SourceT, TargetT);
-    let futex_word: *const TargetT = futex_word as *const SourceT as *const TargetT;
-
-    nix::errno::Errno::result(unsafe {
-        libc::syscall(
-            libc::SYS_futex,
-            futex_word,
-            libc::FUTEX_WAKE,
-            1,
-            core::ptr::null() as *const libc::timespec,
-            core::ptr::null_mut() as *mut u32,
-            0u32,
-        )
-    })?;
-    Ok(())
-}
-#[cfg(loom)]
-pub fn futex_wake(futex_word: &AtomicU32) -> nix::Result<()> {
-    let hashmap = FUTEXES.lock().unwrap();
-    let Some(condvar) = hashmap.get(&(futex_word as *const _ as usize)) else {
-        return Ok(());
-    };
-    condvar.notify_one();
-    Ok(())
+    // In "production" we use linux_syscall to avoid going through libc, and to
+    // avoid touching libc's `errno` in particular.
+    #[cfg(all(not(loom), not(miri)))]
+    {
+        use linux_syscall::Result;
+
+        // The kernel just checks for equality of the value at `futex_word`,
+        // which it interprets as `* const u32`.
+        //
+        // This is safe since `AtomicU32` ["has the same in-memory
+        // representation as the underlying integer type,
+        // u32"](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU32.html#).
+        //
+        // TODO: Consider using
+        // [`as_mut_ptr`](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU32.html#method.as_mut_ptr)
+        // here once it's stabilized.
+        type SourceT = AtomicU32;
+        type TargetT = u32;
+        let futex_word: &SourceT = futex_word;
+        static_assertions::assert_eq_size!(SourceT, TargetT);
+        static_assertions::assert_eq_align!(SourceT, TargetT);
+        let futex_word: *const TargetT = futex_word as *const SourceT as *const TargetT;
+
+        let raw_res = unsafe {
+            linux_syscall::syscall!(
+                linux_syscall::SYS_futex,
+                futex_word,
+                libc::FUTEX_WAKE,
+                1,
+                core::ptr::null() as *const libc::timespec,
+                core::ptr::null_mut() as *mut u32,
+                0u32,
+            )
+        };
+        raw_res
+            .check()
+            .map_err(|e| nix::errno::Errno::from_i32(e.into()))
+    }
+    // Miri understands futex syscalls made through the libc wrapper.
+    // It's ok to depend on libc for miri testing.
+    #[cfg(miri)]
+    {
+        // The kernel just checks for equality of the value at `futex_word`,
+        // which it interprets as `* const u32`.
```

Comment:
It would be nice to be able to dedup this logic for "miri" and "not-miri" so that we don't end up with bugs in only one implementation, but I don't really see a nice way of doing that. Maybe we could write wrapper syscall functions that have the same signature, and then we could assign that to a function pointer to call either using the same arguments. But I think this is fine for this PR, and maybe futex is the only syscall where we need to deal with this problem.